debian-bo-master/bo/source/net/fwtk-1.3/doc/man/login-sh.8
.\" Marcus J. Ranum, 1993
.\" Copyright (C) Trusted Information Systems, Inc.
.\" All rights reserved
.TH LOGIN-SH 8 "2 October 1993"
.SH NAME
login-sh - authenticating login shell
.SH SYNOPSIS
.B login-sh
\[invoked from /bin/login\]
.SH DESCRIPTION
.IX  "login-sh"  ""  "\(em authenticating login shell"
.LP
.B login-sh
provides a simple interface to the authentication service for
login by replacing the user's login shell with a "wrapper"
that requires them to authenticate first, then executes the
real login shell.
.B login-sh
may be used in conjunction with or as a replacement for passwords
in the password file
.I /etc/passwd.
The user's actual login shell information is stored in an external
file.
.LP
Note that
.B login-sh
runs as the user, with their permissions. This is attractive, as
it separates the authentication policy from the permissions granting
policy (/bin/login).
.SH OPTIONS
.LP
.B login-sh
reads its configuration rules and permissions
information from the firewall configuration
table
.B netperm-table,
retrieving all rules specified for "login-sh".
The following configuration rules are recognized:
.TP
.B authserver address port
specifies the network address and service port of the authentication
server to use. (See
.B authsrv)
.TP
.B shellfile pathname
specifies a file that contains information about users' login shells.
The shell configuration file. Empty lines and lines with a '#' as the
first character are discarded or treated as comments. The format of
the shell configuration file is a list of entries, one per line, in
the form of:
.sp
.nf
.na
userid  executable parameter-0 [parameter-1] [parameter-n]
.ad
.fi
.sp
The first three values must be defined. The
.I userid
field matches the login name of the user invoking
.B login-sh
from the
.B /etc/passwd
file. The second field should specify the executable
pathname of the program to run after authentication is
completed. The third and remaining fields are parameters to
pass to the executable program, starting at parameter zero.
Many command interpreters check the name of parameter zero
.I (argv[0])
to determine if they are a login shell. For such command
interpreters, one should be careful to define them with the
expected form, typically a leading dash '-'.
.SH INSTALLATION
.LP
To install
.B login-sh
first place the executable in a system area, and define the
.B shellfile
and
.B authserver
options in
.B netperm-table.
Systems that are using
.B login-sh
should have all programs that permit users to change their login
shells disabled, or should have the setuid bit stripped.
.LP
Users password file entries should resemble:
.sp
.nf
.na
mjr::100:10:Marcus J Ranum:/home/mjr:/usr/local/etc/login-sh
.fi
.ad
.sp
with a sample
.B shellfile
entry for
.I mjr
that resembles:
.sp
.nf
.na
mjr	/usr/bin/ksh   -ksh
.fi
.ad
.sp
Note in the example, that the pathname (/usr/bin/ksh) and the first
parameter for the program ("-ksh") are different. There must be a
minimum of two parameters for each login shell that is defined.
.LP
Users wishing both password authentication and secondary authentication
may set passwords on their entries in
.B /etc/passwd
as well as using
.B login-sh.
.SH FILES
.PD 0
.TP 20
.BR netperm-table
.BR shellfile
.BR /etc/passwd
.SH SEE ALSO
.BR netperm-table (5)
.BR authsrv (8)
.SH BUGS
.LP
Some systems may require the administrator to add login-sh to
.B /etc/shells
as programs like
.B ftpd
check for a valid user shell before permitting access.


debian-bo-master/bo/source/net/fwtk-1.3/tools/server/login-sh/login-sh.c
/*-
 * Copyright (c) 1993, Trusted Information Systems, Incorporated
 * All rights reserved.
 *
 * Redistribution and use are governed by the terms detailed in the
 * license document ("LICENSE") included with the toolkit.
 */

/*
 *	Author: Marcus J. Ranum, Trusted Information Systems, Inc.
 */
static	char	RcsId[] = "Header: login-sh.c,v 1.1 93/10/20 11:17:17 mjr rel ";


#include	<stdio.h>
#include	<pwd.h>
#include	<ctype.h>
#include	<syslog.h>

extern	char	*index();
extern	char	*rindex();
extern	char	*getpass();

#include	"firewall.h"


Cfg		*confp;

main(ac,av)
int	ac;
char	*av[];
{
	Cfg		*cp;
	FILE		*shelf;
	char		*shf;
	struct	passwd	*pw;
	char		buf[BUFSIZ];
	char		tbuf[BUFSIZ];
	char		*tokav[512];
	int		tokac;
	int		logging = 0;

#ifndef	LOG_DAEMON
	openlog("login-sh",LOG_PID);
#else
	openlog("login-sh",LOG_PID|LOG_NDELAY,LFAC);
#endif

	/* if av[0] = log-login-sh then log */
	if((shf = rindex(av[0],'/')) != (char *)0)
		*shf = '\0';
	else
		shf = av[0];
	if(!strcmp(shf,"log-login-sh"))
		logging = 1;

	if((confp = cfg_read("login-sh")) == (Cfg *)-1) {
		fprintf(stderr,"Cannot read config.\n");
		exit(1);
	}

	if((cp = cfg_get("shellfile",confp)) == (Cfg *)0) {
		fprintf(stderr,"No \"shellfile\" defined.\n");
		exit(1);
	}
	if(cp->argc != 1) {
		fprintf(stderr,"shellfile may only have one parameter, line %d\n",cp->ln);
		exit(1);
	}

	if((pw = getpwuid(getuid())) == (struct passwd *)0) {
		fprintf(stderr,"No password entry for uid %d\n",getuid());
		exit(1);
	}

	/* open shells file */
	shf = cp->argv[0];
	if((shelf = fopen(shf,"r")) == (FILE *)0) {
		perror(shf);
		exit(1);
	}

	/* look for user in shells file */
	while(1) {
		if(fgets(buf,sizeof(buf),shelf) == (char *)0) {
			fprintf(stderr,"%s is not in the shells file\n",pw->pw_name);
			exit(1);
		}

		if(buf[0] == '\n' || buf[0] == '#')
			continue;
		tokac = enargv(buf,tokav,512,tbuf,sizeof(tbuf));
		if(tokac <= 0)
			continue;

		if(!strcmp(pw->pw_name,tokav[0]))
			break;
	}
	fclose(shelf);

	if(tokac < 3) {
		fprintf(stderr,"Not enough shell arguments for %s\n",pw->pw_name);
		exit(1);
	}

	/* found user, now authenticate them */
	if(getauth(pw->pw_name))
		exit(1);

	/* set the SHELL environment variable to the real shell */
	(void)setenv("SHELL",tokav[1],1);

	/* authenticated them OK, now invoke the login shell */
	if(logging)
		syslog(LLEV,"securityalert: authenticate user=%s",pw->pw_name);
	execv(tokav[1],&tokav[2]);
	perror(tokav[1]);
	exit(1);
}





getauth(nam)
char	*nam;
{
	char		cbuf[BUFSIZ];
	char		pbuf[512];
	int		bad = 5;
	char		*p;

	/* open connection to auth server */
	if(auth_open(confp)) {
		fprintf(stderr,"Cannot connect to authentication server");
		return(1);
	}

	/* get welcome message from auth server */
	if(auth_recv(cbuf,sizeof(cbuf)))
		goto lostconn;
	if(strncmp(cbuf,"Authsrv ready",13)) {
		fprintf(stderr,"%s\n",cbuf);
		auth_close();
		return(1);
	}

	while(bad--) {
		if(strlen(nam) + 100 > sizeof(cbuf))
			sprintf(cbuf,"authorize %s",nam);
		else
			sprintf(cbuf,"authorize %s 'login-sh'",nam);

		if(auth_send(cbuf))
			goto lostconn;
		if(auth_recv(cbuf,sizeof(cbuf)))
			goto lostconn;

		if(!strncmp(cbuf,"challenge ",10)) {
			fprintf(stderr,"%s",&cbuf[10]);
			if(fgets(pbuf,sizeof(pbuf),stdin) == (char *)0)
				return(1);
			if((p = index(pbuf,'\n')) != (char *)0)
				*p = '\0';
			p = pbuf;
		} else
		if(!strncmp(cbuf,"password",8)) {
			p = getpass("Password: ");
		} else {
			fprintf(stderr,"%s\n",cbuf);
			return(1);
		}

		if(strlen(p) > 64) {
			fprintf(stderr,"Response too long\n");
			continue;
		}

		/* send response */
		sprintf(cbuf,"response '%s'",p);
		if(auth_send(cbuf))
			goto lostconn;
		if(auth_recv(cbuf,sizeof(cbuf)))
			goto lostconn;

		/* no-go */
		if(strncmp(cbuf,"ok",2)) {
			fprintf(stderr,"%s\n",cbuf);
			continue;
		}

		/* go */
		auth_close();
		if(cbuf[2] != '\0')
			fprintf(stderr,"%s\n",&cbuf[2]);
		return(0);
	}
	auth_close();
	fprintf(stderr,"Too many failed logins\n");
	return(1);

lostconn:
	auth_close();
	fprintf(stderr,"Lost connection to auth server\n");
	return(1);
}
